### PR TITLE
[Example] Sparse Graph Transformer

### LINKED ISSUE #5056: [Sparse] Create a transformer example in Sparse API
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by rudongyu**
## Description
close #5056 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sparse/graph_transformer.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        )
+
+
+if __name__ == "__main__":
+    # If CUDA is available, use GPU to accelerate the training, use CPU
+    # otherwise.
+    dev = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+    # Load graph from the existing dataset.
+    dataset = CoraGraphDataset()
```
Comment (jermainewang) [inline examples/sparse/graph_transformer.py:145]: Maybe using a graph prediction dataset to be more real?

File: examples/sparse/graph_transformer.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        return self.batchnorm2(h)
+
+
+class GTModel(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        hidden_size=64,
+        pos_enc_size=2,
```
Comment (mufeili) [inline examples/sparse/graph_transformer.py:77]: Does this have to be the same as in `laplacian_pe`? If so, consider passing it to both `laplacian_pe` and `GTModel` instead.

File: examples/sparse/graph_transformer.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        attn = attn.softmax()
+        out = bspmm(attn, v.permute(0, 2, 1)).permute(0, 2, 1)
+
+        return self.out_proj(out.reshape(N, -1))
+
+
+class GTLayer(nn.Module):
+    """Graph Transformer Layer"""
+
+    def __init__(self, hidden_size=80, num_heads=8) -> None:
```
Comment (mufeili) [inline examples/sparse/graph_transformer.py:51]:
```suggestion
    def __init__(self, hidden_size=80, num_heads=8):
```

Better to be consistent, either include that for all modules or the other way

File: examples/sparse/graph_transformer.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        # Backward.
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+        # Compute prediction.
+        pred = logits.argmax(dim=1)
+
+        # Evaluate the prediction.
+        val_acc, test_acc = evaluate(g, pred)
```
Comment (mufeili) [inline examples/sparse/graph_transformer.py:132]: Missing `model.eval()`

**Review by mufeili — CHANGES_REQUESTED** [review]
Minor comments

File: examples/sparse/graph_transformer.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+(https://arxiv.org/abs/2012.09699)
+"""
+
+import dgl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import bspmm, create_from_coo, mock_bsddmm as bsddmm
```
Comment (mufeili) [inline examples/sparse/graph_transformer.py:12]:
```suggestion
from dgl.sparse import bspmm, create_from_coo, mock_bsddmm as bsddmm
```
with #5088 merged

**Review by jermainewang — APPROVED** [review]
Nice!

**Comment by jermainewang** [comment]
@rudongyu How long is the training and what accuracy does it get?

**Comment by rudongyu** [comment]
> @rudongyu How long is the training and what accuracy does it get?

~75 secs for positional encoding computation and 5mins for training (50 epochs, Tesla T4).
Performance (ROC AUC): Val-79.35, Test-74.77

File: examples/sparse/graph_transformer.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            nn.ReLU(),
+            nn.Linear(hidden_size // 2, hidden_size // 4),
+            nn.ReLU(),
+            nn.Linear(hidden_size // 4, out_size),
+        )
+
+    def forward(self, g, X, pos_enc):
+        src, dst = g.edges()
+        N = g.num_nodes()
+        A = dglsp.create_from_coo(dst, src, shape=(N, N))
```
Comment (mufeili) [inline examples/sparse/graph_transformer.py:103]:
```suggestion
        A = dglsp.from_coo(dst, src, shape=(N, N))
```
with #5093 merged

**Review by mufeili — APPROVED** [review]
